aptos-move/framework/aptos-framework/sources/keyless_account.move

```diff
@@ -98,81 +111,136 @@ module aptos_framework::keyless_account {
+        if (config_buffer::does_exist<Groth16VerificationKey>()) {
+            let vk = config_buffer::extract();
+            if (exists<Groth16VerificationKey>(@aptos_framework)) {
+                *borrow_global_mut<Groth16VerificationKey>(@aptos_framework) = vk;
+            } else {
+                move_to(fx, vk);
+            }
         };
 
-        move_to(fx, config);
+        if(config_buffer::does_exist<Configuration>()) {
+            let config = config_buffer::extract();
+            if (exists<Configuration>(@aptos_framework)) {
+                *borrow_global_mut<Configuration>(@aptos_framework) = config;
+            } else {
+                move_to(fx, config);
+            }
+        };
     }
 
-    // Convenience method to [un]set the training wheels PK, only callable via governance proposal.
-    // WARNING: If a malicious key is set, this would lead to stolen funds.
+    /// Convenience method to queue up a change to the training wheels PK. The change will only be effective after
+    /// reconfiguration. Only callable via governance proposal.
+    ///
+    /// WARNING: If a malicious key is set, this *could* lead to stolen funds.
     public fun update_training_wheels(fx: &signer, pk: Option<vector<u8>>) acquires Configuration {
         system_addresses::assert_aptos_framework(fx);
+
+        // If a PK is being set, validate it first.
         if (option::is_some(&pk)) {
-            assert!(vector::length(option::borrow(&pk)) == 32, E_TRAINING_WHEELS_PK_WRONG_SIZE)
+            let bytes = *option::borrow(&pk);
+            let vpk = ed25519::new_validated_public_key_from_bytes(bytes);
+            assert!(option::is_some(&vpk), E_TRAINING_WHEELS_PK_WRONG_SIZE)
+        };
+
+        let config = if (config_buffer::does_exist<Configuration>()) {
+            config_buffer::extract<Configuration>()
+        } else {
+            *borrow_global<Configuration>(signer::address_of(fx))
         };
 
-        let config = borrow_global_mut<Configuration>(signer::address_of(fx));
         config.training_wheels_pubkey = pk;
+
+        update_configuration(fx, config);
     }
 
-    // Convenience method to set the max expiration horizon, only callable via governance proposal.
+    /// Convenience method to queues up a change to the max expiration horizon. The change will only be effective after
+    /// reconfiguration. Only callable via governance proposal.
     public fun update_max_exp_horizon(fx: &signer, max_exp_horizon_secs: u64) acquires Configuration {
         system_addresses::assert_aptos_framework(fx);
 
-        let config = borrow_global_mut<Configuration>(signer::address_of(fx));
+        let config = if (config_buffer::does_exist<Configuration>()) {
+            config_buffer::extract<Configuration>()
+        } else {
+            *borrow_global<Configuration>(signer::address_of(fx))
+        };
+
         config.max_exp_horizon_secs = max_exp_horizon_secs;
+
+        update_configuration(fx, config);
```

Comment:
just a bookmark

Reply:
this is not used in the cache right?

Reply:
not used from my understanding

Reply:
it's a bit confusing to have both direct update and on epoch update

Reply:
There is no direct update. `update_configuration` queues up the update in the config buffer...